### ISSUE TITLE
periodic assertion expired error for firefox marketplace

### ISSUE THREAD

**Issue by jedp** [mozilla/persona#3033]
@kumar303 reported this in bugzilla [bug 843795](https://bugzilla.mozilla.org/show_bug.cgi?id=843795)

(n.b., marketplace is currently using the integration server native-persona.org, which is where the b2g branch is deployed)

Kumar writes:

Every couple days (sometimes weeks) I can't log in to Marketplace on B2G because of an "assertion expired" error.

When I rebuild my profile the error goes away.

I've seen it on many different versions: localhost, dev, stage, etc. Here is the latest failure on stage so you can unpack the assertion and check it out. Maybe something is cached client side that shouldn't be?

```
Feb 21 13:05:45 web6.stage.addons.phx1.mozilla.com: [][] django_browserid.base:INFO Verification URL: https://native-persona.org/verify :/data/www/addons.allizom.org/venv/src/django-browserid/django_browserid/base.py:109
Feb 21 13:05:45 web6.stage.addons.phx1.mozilla.com: [][] django_browserid.base:ERROR BrowserID verification failure. Response: {u'status': u'failure', u'reason': u'assertion has expired'} Audience: https://marketplace.allizom.org :/data/www/addons.allizom.org/venv/src/django-browserid/django_browserid/base.py:121
Feb 21 13:05:45 web6.stage.addons.phx1.mozilla.com: [][] django_browserid.base:ERROR BID assert: eyJhbGciOiJSUzI1NiJ9.eyJwdWJsaWMta2V5Ijp7ImFsZ29yaXRobSI6IkRTIiwieSI6IjcyYjEyMGY3ZDljMzIxMzFmOWUzY2I5NDU3ODc4MzI4NGM3YzhmYmJlMjkxYzVjMzM2OGRkNjgyMzA2MjM0MWRlYjAxYzEyMzIxNjM0OWM4ODVhNDM0NGY4Mzg1ZGViOTQzMWY4ZTg1NzA2MmI3Y2UxODEyNjQyZDJjODFiYTZkNDQxNzE1NjU4ODQyNjUyYjllZGM5NTY0NDYyNjdiN2ExN2JiNzZjNWRlNGQ3NTU1NzZlMTlhMmUyMDZlNTczZDMxNzBjNjg2NjQyNzAxMzY2NjQ2NDRmODc4OTRiNGY2MWFlNmM4MDY5MTM4MWE4NjU1ZWIzOTFlOTkyNjhkNjkiLCJwIjoiZmY2MDA0ODNkYjZhYmZjNWI0NWVhYjc4NTk0YjM1MzNkNTUwZDlmMWJmMmE5OTJhN2E4ZGFhNmRjMzRmODA0NWFkNGU2ZTBjNDI5ZDMzNGVlZWFhZWZkN2UyM2Q0ODEwYmUwMGU0Y2MxNDkyY2JhMzI1YmE4MWZmMmQ1YTViMzA1YThkMTdlYjNiZjRhMDZhMzQ5ZDM5MmUwMGQzMjk3NDRhNTE3OTM4MDM0NGU4MmExOGM0NzkzMzQzOGY4OTFlMjJhZWVmODEyZDY5YzhmNzVlMzI2Y2I3MGVhMDAwYzNmNzc2ZGZkYmQ2MDQ2MzhjMmVmNzE3ZmMyNmQwMmUxNyIsInEiOiJlMjFlMDRmOTExZDFlZDc5OTEwMDhlY2FhYjNiZjc3NTk4NDMwOWMzIiwiZyI6ImM1MmE0YTBmZjNiN2U2MWZkZjE4NjdjZTg0MTM4MzY5YTYxNTRmNGFmYTkyOTY2ZTNjODI3ZTI1Y2ZhNmNmNTA4YjkwZTVkZTQxOWUxMzM3ZTA3YTJlOWUyYTNjZDVkZWE3MDRkMTc1ZjhlYmY2YWYzOTdkNjllMTEwYjk2YWZiMTdjN2EwMzI1OTMyOWU0ODI5YjBkMDNiYmM3ODk2YjE1YjRhZGU1M2UxMzA4NThjYzM0ZDk2MjY5YWE4OTA0MWY0MDkxMzZjNzI0MmEzODg5NWM5ZDViY2NhZDRmMzg5YWYxZDdhNGJkMTM5OGJkMDcyZGZmYTg5NjIzMzM5N2EifSwicHJpbmNpcGFsIjp7InVudmVyaWZpZWQtZW1haWwiOiJrdW1hci5tY21pbGxhbitxMkBnbWFpbC5jb20ifSwiaWF0IjoxMzYxNDcxMDk0NzA5LCJleHAiOjEzNjE0NzQ2OTQ3MDksImlzcyI6ImIyZzJwYWMucGVyc29uYXRlc3Qub3JnIn0.BLUi8HmIxX-zLlCuj6apipnBt8iI5FcZ8joqVa0Y9HrpGBmFZI_318AdHZuHm7Jw0tcldJ268KNmWflrt3wF2eiNcAKlYFh1-HZ9-B88sndYgTYNx03g0AC62Az-RGwuPiVCOzZp_QdiqQMZGvKsQbApYNZKH3mVeyT5oonb1OC8eirZ3wg25oZhqpweOkQItArXaw4YwYQBUiJ9Twv9_c8fmVtnI_Q3sAXOTANitWpQyoPqhGHcNHX2Us0Lqco7oXQlf0W4H51Uqq2JEGsyH5Y0wV8yJzJe8_KziMHoOcevhjVVowAYKoMoVlPgMKCD-3FfQvAOAowyGAUNNGOB4w~eyJhbGciOiJEUzEyOCJ9.eyJleHAiOjEzNjE0ODA4NjM3MjcsImF1ZCI6Imh0dHBzOi8vbWFya2V0cGxhY2UuYWxsaXpvbS5vcmcifQ.ZLQH_YPbB6AKQRM5Tffc-MfQ0AkMrisQxrhiCHkGgySSSjd909bTTQ
```

**Comment by seanmonstar** [comment]
How do I check the configuration on native-persona.org?

The b2g.json config file needs to be used, and it should have the expiration set to 10 years out: https://github.com/mozilla/browserid/blob/b2g/config/b2g.json

**Comment by jedp** [comment]
@seanmonstar thanks! it's an awsbox: `ssh app@native-persona.org`

**Comment by jrgm** [comment]
@jedp - that assertion is for issuer b2g2pac.personatest.org ?

**Comment by jrgm** [comment]
@jedp - my point about b2g2pac is that the DNS for that is currently broken (a lookup returns a CNAME with an IP and not another hostname). So an http request for the public key can never work. (Sorry, if I'm off topic).

**Comment by seanmonstar** [comment]
I've distracted us by bringing up that config value: that affects cookie length on native-persona.org itself.

This is an issue with the assertion being expired by the time it's being used. Is it from time shift on the awsbox? @jrgm makes a good point also: why is the issuer from b2g2pac?

**Comment by jrgm** [comment]
I looked at `ntpq -p` on native-persona.org and we appear to be synchronized at the moment.

**Comment by jedp** [comment]
`b2g2pac` is the test issuer for the `forceIssuer` argument, which is something marketplace uses to have assertions issued on behalf of a certain origin that is not actually a primary (see [bug 811014](https://bugzilla.mozilla.org/show_bug.cgi?id=811014)).  That will presumably be a more marketplacey url as things go into production.  Note that the lists of issuers is a whitelist in the `config.json` file.

I don't _think_ it should matter that the DNS for b2g2pac is borked, because with `forceIssuer`, part of the point is not to do primary discovery.

**Comment by clouserw** [comment]
What news on this? :)

**Comment by seanmonstar** [comment]
What's new is that someone on [signin] needs to grab this. I'm far too busy on [picl], and have no idea where to look further.

/cc @lloyd

**Comment by lloyd** [comment]
clouserw | kumar: what device or environment are you seeing this on?

My current best theory is that device clock skew is causing this.  Because the persona bits are long-lived on the device, there is a large window of vulnerability where skew could occur.

based on your answer I have a couple different ideas about what we can do...

**Comment by lloyd** [comment]
also, how long has the device/environment been running at the time you see the error?

also, once this error occurs, is it consistent?  like you'll never see the situation "fix itself" without a profile rebuild?

If we can confirm that your devices clock is off at the time that this occurs I think the solution is rather simple on our side.

**Comment by lloyd** [comment]
finally!  I know there were some issues with clock stability at MWC.  I believe the geeksphone was the device affected by this.  This is a wild guess though, just searching for correlation.

**Comment by kumar303** [comment]
HI lloyd. Pretty much anyone who tests the marketplace has run into this. We were getting it on IRC all the time but everyone knows the workaround now. 

I see it a lot on desktop B2G but others are using Unagi. I don't know of any case where it has fixed itself. In all cases, when a user re-flashes, does a factory reset, or builds a clean desktop profile, it is always fixed.

I can't say for sure how long it takes. I usually re-build my desktop b2g every couple of days. If I haven't rebuilt in a while, then I'll hit the error. So... it takes at least a couple days, I think?

**Comment by lloyd** [comment]
all of this is making me believe more strongly that this is caused by client side clock skew of > 2m from when the first time a user interacts with persona to the time they see the problem.

can we send out an APB that the next time someone sees this they check their system clock and report the delta of current system time (phone or computer) to actual time?

In terms of resolution, I belive the same fix that @jedp is working on to address resident memory concerns will also fix this problem.

@jedp, can you confirm that your fix will make it so there is a max delta between the time the dialog is spawned to when an assertion is issued of about 5-10mins?  Also, what is the timing of that fix?

**Comment by lloyd** [comment]
@jedp, assigning to you, please push this back at me if duty calls you to other places :beers:

**Comment by kumar303** [comment]
I can send an email to our dev list. Do you mean to have the person look at the clock in desktop B2G and see if there is a delta from their computer's clock?

**Comment by jedp** [comment]
First thing is for me to have the verifier error report the current time on device, as well as the assertion issue and expiration times.  And as Lloyd says, it seems not improbable that the memshrink patch will solve the problem.  I am almost ready to land that patch, pending one issue I'm trying to resolve regarding calls to nav.id when the device is offline.

**Comment by jedp** [comment]
Sounds like this is biting again: http://irclog.gr/#show/irc.mozilla.org/identity/245406

Lloyd dug in a bit on this with me.  He noted that the expired assertions all appear to come from the hidden communication_iframe (not the sign_in dialog).  Starting at Mon Apr 22 16:40:45 PDT 2013, expired assertions will be printed in the verifier-metrics.json log.  Trap has been set.  Will look for captured prey tomorrow.

**Comment by jedp** [comment]
Worked with @lloyd on this yesterday, and pushed a fix (train 2013.3.29 branch had the hotfix).  Looking at the logs now, I see there were no new failures in the last two workdays in Spain.  So I think this issue is fixed.

Closing.
